src/internal_telemetry/allocations/mod.rs

```diff
@@ -22,14 +22,25 @@ pub(crate) use self::allocator::{
 
 use crossbeam_utils::CachePadded;
 
-const NUM_GROUPS: usize = 256;
+const NUM_GROUPS: usize = 128;
 /// These arrays represent the allocations and deallocations for each group.
 /// We pad each Atomic to reduce false sharing effects.
-static GROUP_MEM_ALLOCS: [CachePadded<AtomicU64>; NUM_GROUPS] =
-    arr![CachePadded::new(AtomicU64::new(0)); 256];
-static GROUP_MEM_DEALLOCS: [CachePadded<AtomicU64>; NUM_GROUPS] =
-    arr![CachePadded::new(AtomicU64::new(0)); 256];
-
+static GROUP_MEM_STATS: [[CachePadded<AtomicI64>; NUM_GROUPS]; 8] = [
+    arr![CachePadded::new(AtomicI64::new(0)); 128],
+    arr![CachePadded::new(AtomicI64::new(0)); 128],
+    arr![CachePadded::new(AtomicI64::new(0)); 128],
+    arr![CachePadded::new(AtomicI64::new(0)); 128],
+    arr![CachePadded::new(AtomicI64::new(0)); 128],
+    arr![CachePadded::new(AtomicI64::new(0)); 128],
+    arr![CachePadded::new(AtomicI64::new(0)); 128],
+    arr![CachePadded::new(AtomicI64::new(0)); 128],
+];
+
+static THREAD_COUNTER: AtomicUsize = AtomicUsize::new(0);
+
+thread_local! {
+    static THREAD_ID: AtomicUsize = AtomicUsize::new(THREAD_COUNTER.fetch_add(1, Ordering::SeqCst) + 1);
```

Comment:
Two things:

1. We should leave a TODO comment here that `std::thread::ThreadId::as_u64` will eventually be able to provide this for us when it's stabilized.
2. We should be able to avoid actually needing an atomic for the thread local itself: the value never changes once initialized.